core/src/main/java/org/apache/calcite/sql/SqlUpdate.java

```diff
@@ -160,12 +166,104 @@ public void setSourceSelect(SqlSelect sourceSelect) {
         writer.startList(SqlWriter.FrameTypeEnum.SELECT, "UPDATE", "");
     final int opLeft = getOperator().getLeftPrec();
     final int opRight = getOperator().getRightPrec();
+
+    List<SqlNode> sources = new ArrayList<>();
+
+    if (sourceSelect != null && sourceSelect.from != null) {
+      SqlJoin join = getJoinFromSourceSelect();
+      if (join != null) {
+        setTargetAndSources(join, sources);
+      }
+    }
+
     targetTable.unparse(writer, opLeft, opRight);
+
     SqlIdentifier alias = this.alias;
     if (alias != null) {
       writer.keyword("AS");
       alias.unparse(writer, opLeft, opRight);
     }
+
+    unparseSet(writer, opLeft, opRight);
+    if (!sources.isEmpty()) {
+      unparseUpdateSources(sources, writer, opLeft, opRight);
+    }
+    unparseUpdateCondition(writer, opLeft, opRight);
+
+    writer.endList(frame);
+  }
+
+  private SqlIdentifier getFromAlias() {
```

Comment:
Can the method name be changed to something more elaborate like "getAliasForFromClause"

Reply:
Done